### ISSUE TITLE
Public type should be ignored, if returned in pub fun; type usage in pattern matching not working

### ISSUE THREAD

**Issue by raleng** [darky/cleam#1]
Hey there,

I just tested cleam on a project of mine and noticed a false positives regarding public types.

```gleam
// foo.gleam
pub type Foo {
  Foo
}

pub fn foo() -> Foo {
  Foo
}
```

```gleam
// main.gleam
import foo

pub fn main() {
  foo.foo()
}
```

The type must be public because it's used in a public function.

Even if I reference the variant, cleam still claims the type is not used:

```gleam
// main.gleam
import foo

pub fn main() {
  case foo.foo() {
    foo.Foo -> ""
  }
}
```

The only way to make cleam happy is by importing the constructor directly.

```gleam
// main.gleam
import foo.{Foo}

pub fn main() {
  case foo.foo() {
    Foo -> ""
  }
}
```

A typical situation where this comes up for me, is when the type is custom error type. I don't want to make it opaque, because I'd like other modules to be able to pattern match over the error variants to construct whatever response is relevant for the context of that module.

Cheers!

**Comment by darky** [comment]
Hi, thanks for feedback!

```gleam
// main.gleam
import foo

pub fn main() {
  case foo.foo() {
    foo.Foo -> ""
  }
}
```

☝️ Pattern matching type usage fixed in `1.4.1`

```gleam
// foo.gleam
pub type Foo {
  Foo
}

pub fn foo() -> Foo {
  Foo
}
```

☝️ Implicit usage of type via public function harder to fix. But I will try to fix it near future :)

**Comment by darky** [comment]
```gleam
// foo.gleam
pub type Foo {
  Foo
}

pub fn foo() -> Foo {
  Foo
}
```

☝️ Ok, now type will be ignored, if it's used as return type of public function in explicit way
Please upgrade to `1.5.0`
